### DISCUSSION TITLE
Compile method does not work with List

### DISCUSSION THREAD

**Discussion by 9en9i** [sqlalchemy/sqlalchemy#11856] [Usage Questions] [answered]
Hi, I am doing a project migration from psycopg2 to psycopg 3. I can often see this line in my code:
`func.coalesce(field, ‘{}’)`
But with psycopg 3 it doesn't work, it does:
`func.coalesce(field, [])`
The problem is that then the `compile` method stops working.
For example, I get the error `sqlalchemy.exc.CompileError: No literal value renderer is available for literal value ‘[]’ with datatype NULL` in this code:
```python
from sqlalchemy import select, func
from sqlalchemy.dialects import postgresql

stmt = select(func.coalesce(func.sum(1), []))
print(stmt.compile(
    compile_kwargs={‘literal_binds’: True},
    dialect=postgresql.dialect(),
))
```

I suspect this may still be the case with other types such as dict. Can you give me a workaround?

**Comment by CaselIT** [comment]
Hi,

First alternative is to not literal binds. There should be no need to use it outside debug. Note that the same issue happens in any postgresql dialect, it doesn't matter if it's psycopg2 or 3 used.

Other alternative could be to provide your own subclass of the postgresql dialect that defines how to stringify these types. 

Something like
```py
class _LiteralString(String):
  def literal_processor(self, dialect):
    def process(value):
      # if isinstance(...): # provide custom rendering here
      return str(value)

class _DialectForLiteral(postgresql.dialect):
  colspecs = postgresql.dialect | {
    NullType: _LiteralString,
  }
```

but again I would not use the above outside of debug, since there is the possibility of sql injection with the above code

**Reply by 9en9i** [reply]
Thanks for the reply. Yes, you are right, it also didn't work in psycopg 2. But "{}" doesn't work in psycopg 3, so I have to change it to "[]". Yes, I use it purely for debugging purposes. In any case, I'll try what you suggested.

**Reply by CaselIT** [reply]
I you aren't planning on running that statement (you shouldn't) you can take advantage of the fact that psycopg2/3 use py-format for bind arguments, so you can just do
```py
compiled = stmt.compile(dialect=postgresql.dialect())
print(compiled.string % compiled.params)
```

**Reply by 9en9i** [reply]
The main reason for using it is just to see the finished sql query and possible execution for debugging.

**Reply by 9en9i** [reply]
I also tried your example with _LiteralString, maybe I misunderstood and I need some _LiteralArray? Because it didn't work for me now

**Reply by CaselIT** [reply]
> The main reason for using it is just to see the finished sql query and possible execution for debugging.

using `%` should give you that.



> I also tried your example with _LiteralString, maybe I misunderstood and I need some _LiteralArray? Because it didn't work for me now

You need to override all the types you want to support in the colspecs dict. It may be that you have to override the array type. Also note that you should then use that dialect instead of the pg one
```py
print(stmt.compile(
    compile_kwargs={‘literal_binds’: True},
    dialect=_DialectForLiteral(),
))
```

**Reply by 9en9i** [reply]
No, using `%` is not appropriate.  I'm getting
`SELECT coalesce(sum(1), []) AS coalesce_1`. Here `[]` is a invalid literal for the array

**Reply by CaselIT** [reply]
you may need to customize the output of the `_LiteralString` too then if you want a runnable sql string then

**Reply by 9en9i** [reply, marked as answer]
I got code like this:
```python
class _LiteralString(String):
    @override
    def literal_processor(self, dialect: postgresql.dialect) -> Any:
        def process(value: object) -> str:
            if isinstance(value, list):
                return "'{}'".format(str(value).replace("[", "{").replace("]", "}"))
            return str(value)

        return process


class _DialectForLiteral(postgresql.dialect):
    colspecs = postgresql.dialect.colspecs | {
        NullType: _LiteralString,
    }
```

**Reply by 9en9i** [reply]
Maybe I can open a PR to add literal type support for `[]` and others? I will only need you to point out where I need to pay attention (I mean how it's supposed to work).

**Reply by CaselIT** [reply]
In general sqlalchemy tries to be quite conservative in this case since there may be security issues in liberal binding.

I don't think the complexity of having a correct literal renderer for this is worth it
